### ISSUE TITLE
Tune not working with classification models

### ISSUE THREAD

**Issue by trnikon** [ultralytics/ultralytics#4659]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

Evolution

### Bug

It seems like hyperparameter tuning code only works with detection models. With pretrained classification model, training works fine, but when I try to use model.tune() on a standard or custom dataset, I get a series of errors. 

#Examples:
Tuning Caltech101: 
`RuntimeError: Dataset 'caltech101' error ❌ [Errno 2] No such file or directory: '/project/datasets/caltech101/train'`

Tuning custom dataset at the end of epoch 1: 
```
Traceback (most recent call last):
  File "/home/trnikon/project/yolov8.py", line 17, in <module>
    main()
  File "/home/trnikon/project/yolov8.py", line 9, in main
    model.tune(data='/home/trnikon/project/DATASETS/combined', imgsz=128, epochs=30, iterations=300)
  File "/home/trnikon/miniconda3/envs/yolo/lib/python3.10/site-packages/ultralytics/engine/model.py", line 365, in tune
    return Tuner(args=args, _callbacks=self.callbacks)(model=self.model, iterations=iterations)
  File "/home/trnikon/miniconda3/envs/yolo/lib/python3.10/site-packages/ultralytics/engine/tuner.py", line 172, in __call__
    results = model.train(**train_args)
  File "/home/trnikon/miniconda3/envs/yolo/lib/python3.10/site-packages/ultralytics/engine/model.py", line 341, in train
    self.trainer.train()
  File "/home/trnikon/miniconda3/envs/yolo/lib/python3.10/site-packages/ultralytics/engine/trainer.py", line 196, in train
    self._do_train(world_size)
  File "/home/trnikon/miniconda3/envs/yolo/lib/python3.10/site-packages/ultralytics/engine/trainer.py", line 392, in _do_train
    self.metrics, self.fitness = self.validate()
  File "/home/trnikon/miniconda3/envs/yolo/lib/python3.10/site-packages/ultralytics/engine/trainer.py", line 498, in validate
    metrics = self.validator(self)
  File "/home/trnikon/miniconda3/envs/yolo/lib/python3.10/site-packages/torch/utils/_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
  File "/home/trnikon/miniconda3/envs/yolo/lib/python3.10/site-packages/ultralytics/engine/validator.py", line 191, in __call__
    stats = self.get_stats()
  File "/home/trnikon/miniconda3/envs/yolo/lib/python3.10/site-packages/ultralytics/models/yolo/detect/val.py", line 146, in get_stats
    self.nt_per_class = np.bincount(stats[-1].astype(int), minlength=self.nc)  # number of targets per class
IndexError: list index out of range
```

### Environment

- YOLO: 8.0.166
- OS: Ubuntu 22.04.3 LTS
- Python: 3.10.12
- GPU: Nvidia GeForce RTX 4090
- CPU: AMD Ryzen 9 7945HX

### Minimal Reproducible Example

from ultralytics import YOLO

model = YOLO('yolov8l-cls.pt')

model.tune(data='caltech101', imgsz=416, epochs=30, iterations=300)

//alternatively with custom dataset
model.tune(data='/path-to-dataset', imgsz=128, epochs=30, iterations=300)

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@trnikon I'm working on Tuner updates and fixes currently in #4656. The following Classification tuning works correctly for me there. Can you try to checkout the branch and see if it resolves your issue?


```python
from ultralytics import YOLO

model = YOLO('yolov8n-cls.pt')
model.tune(data='imagenet10', imgsz=64, epochs=5, iterations=10, val=False, cache=True)
```

**Comment by trnikon** [comment]
@glenn-jocher Currently with the pip package I get `FileNotFoundError: Dataset 'imagenet10' for task=detect not found ❌`. Which branch do I check out?

**Comment by glenn-jocher** [comment]
Branch is called export_fix and this should be merged into a new pip update later today or tomorrow

**Comment by glenn-jocher** [comment]
It turns out the original Tuner PR was merged a bit too soon so I've been debugging and correcting it and now it seems to work well, just like my yolov5 one essentially.

**Comment by trnikon** [comment]
@glenn-jocher Installing from branch (pip install git+https://github.com/ultralytics/ultralytics.git@export_fix) seems to fix it. I will update after tuning ends

**Comment by glenn-jocher** [comment]
@trnikon great news 😃! Your original issue may now be resolved ✅ in PR #4656 and published as `ultralytics 8.0.167` just now. 

To get this update:
- **[Git](https://github.com/ultralytics/ultralytics)** – Run `git pull` from within your `ultralytics/` directory or run `git clone https://github.com/ultralytics/ultralytics` again
- **[Pip](https://pypi.org/project/ultralytics/)** – Update with `pip install -U ultralytics`
- **[Notebooks](https://github.com/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb)** – Check out the updated notebooks  <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/ultralytics/yolov8"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **[Docker](https://hub.docker.com/r/ultralytics/ultralytics)** – Run `sudo docker pull ultralytics/ultralytics:latest` to update your image <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

Thank you for spotting this issue and letting us know. Please confirm if this update fixes the issue for you, and don't hesitate to report any other **issues** you find or **feature requests** you may have. Happy training with YOLOv8 🚀!

**Comment by trnikon** [comment]
I stopped tuning early (ctrl+c) because the number of iterations I put was too high and it was taking too long and I noticed that tuning only saved the last weights, not the best weights. I searched manually for the best training session because I got no reference to it in the console output and `_tune/evolve.csv` had no reference to the trial number and used a different metric than `top1_acc` called `fitness_score` whose top value indicated a different trial. I tried loading the configuration of the best training session from the `args.yaml` file in order to retrain using them but I couldn't find documentation on it and was unsuccessful.

**Comment by glenn-jocher** [comment]
@trnikon hello, thank you for highlighting these areas of concern. I understand that you're facing a few challenges; let's address each one:

1. With regards to tuning saving only the last weights and not the best one, this has been observed and is part of the current design. However, we appreciate your feedback and we'll consider your suggestion for an update to keep the best weights in future releases.

2. It seems like you encountered some confusion with `fitness_score` metric in `evolve.csv. `fitness_score` is a metric used to evaluate and compare configurations during the tuning process. While it's true it isn't directly comparable to `top1_acc`, it indicates the relative performance across different trials.

3. As for loading the model configuration from `args.yaml` and retraining with it, this might be a bit tricky since there's currently no high-level function in the package to perform this straightforwardly. I understand how this could be a point of frustration, and we're working on ways to improve our documentation and functions to make this process more seamless for end-users.

We genuinely appreciate your insightful feedback which helps us improve our project. If you have any more questions or concerns, feel free to ask. Happy training with YOLOv8!

**Comment by glenn-jocher** [comment]
@trnikon hey thanks for your feedback! You're right that we should output a best.yaml on every iteration, not just the final one. 

Fitness varies by task but for Classify tasks is a combination of top1 and top5 accuracy. You should be able to scan your results CSV file and find the highest fitness row, then these are your best hyperparameters. 

I'll make a note to add per-iteration output of best results to the Tuner.

**Comment by glenn-jocher** [comment]
@trnikon great news 😃! Your original issue may now be resolved ✅ in PR #4680. This PR improves Tuner results introspection by printing and saving the best results at the end of every iteration, not just after all iterations are complete. 

This should significantly help see your results progress over time and allow you to cancel at anytime while still retaining all current progress.

To get this update:
- **[Git](https://github.com/ultralytics/ultralytics)** – Run `git pull` from within your `ultralytics/` directory or run `git clone https://github.com/ultralytics/ultralytics` again
- **[Pip](https://pypi.org/project/ultralytics/)** – Update with `pip install -U ultralytics`
- **[Notebooks](https://github.com/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb)** – Check out the updated notebooks  <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/ultralytics/yolov8"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **[Docker](https://hub.docker.com/r/ultralytics/ultralytics)** – Run `sudo docker pull ultralytics/ultralytics:latest` to update your image <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

Thank you for spotting this issue and letting us know. Please confirm if this update fixes the issue for you, and don't hesitate to report any other **issues** you find or **feature requests** you may have. Happy training with YOLOv8 🚀!

**Comment by glenn-jocher** [comment]
Results look like this for me now. BTW if you set `val=False` and `save=False` during tuning then only final epoch results are evaluated and then last.pt and best.pt are the same.

<img width="1385" alt="Screenshot 2023-08-31 at 23 43 28" src="https://github.com/ultralytics/ultralytics/assets/26833433/642e0c6b-0c4c-41a0-92a0-3622bc713021">
